sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TriggerExecutor.scala

```diff
@@ -65,8 +65,22 @@ case class ProcessingTimeExecutor(processingTime: ProcessingTime, clock: Clock =
       s"${intervalMs} milliseconds, but spent ${realElapsedTimeMs} milliseconds")
   }
 
-  /** Return the next multiple of intervalMs */
+  /** Return the next multiple of intervalMs
+   *
+   * e.g. for intervalMs = 100
+   * nextBatchTime(0) = 100
+   * nextBatchTime(1) = 100
+   * ...
+   * nextBatchTime(99) = 100
+   * nextBatchTime(100) = 200
+   * nextBatchTime(101) = 200
+   * ...
+   * nextBatchTime(199) = 200
+   * nextBatchTime(200) = 300
+   *
+   * Note, this way, we'll get nextBatchTime(nextBatchTime(0)) = 200, rather than = 0
+   * */
   def nextBatchTime(now: Long): Long = {
-    (now - 1) / intervalMs * intervalMs + intervalMs
+    now / intervalMs * intervalMs + intervalMs
```

Comment:
When I wrote this method, I was trying to deal with one case: If a batch takes exactly `intervalMs`, we should run the next batch at once instead of sleeping `intervalMs`. This changes will break it.

However, I forgot to handle the case that a batch takes 0ms. How about changing [this line](https://github.com/apache/spark/blob/0513c3ac93e0a25d6eedbafe6c0561e71c92880a/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/TriggerExecutor.scala#L53) to:

```Scala
        if (batchElapsedTimeMs == 0) {
          clock.waitTillTime(intervalMs)
        } else {
          clock.waitTillTime(nextBatchTime(batchEndTimeMs))
        }
```